### PR TITLE
Add `dotted` output format

### PR DESCRIPTION
**Pull request by oschaaf**
This adds `--output-format dotted` as an option. Doing so will make
the client emit a flattened list of dotted strings with `key: value`s.
This simple format is useful for processing output with minimal
dependencies on other libraries like json/proto/etc, and prometheus
also supports it.

Fixes https://github.com/envoyproxy/nighthawk/issues/136

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
@jplevyak this is ready for review

File: source/client/output_collector_impl.cc [hunk 1]
```diff
@@ -122,5 +122,51 @@ std::string YamlOutputCollectorImpl::toString() const {
   return Envoy::MessageUtil::getYamlStringFromMessage(toProto(), true, true);
 }
 
+DottedStringOutputCollectorImpl::DottedStringOutputCollectorImpl(Envoy::TimeSource& time_source,
+                                                                 const Options& options)
+    : OutputCollectorImpl(time_source, options) {}
+
+std::string DottedStringOutputCollectorImpl::toString() const {
```
Comment (htuch) [inline source/client/output_collector_impl.cc]: As per my comment on the other review, can we make this a pure function that takes a Nighthawk proto output and performs the format? That way we can avoid putting this on the wire and allow clients to decide how they want to do the output formatting.

Comment (oschaaf) [reply source/client/output_collector_impl.cc]: So as we are planning to have [a separate CLI](https://github.com/envoyproxy/nighthawk/pull/159) instead, I think we can pursue landing this as-is, correct?

File: source/client/output_collector_impl.cc [hunk 2]
```diff
@@ -44,7 +44,7 @@ std::string ConsoleOutputCollectorImpl::toString() const {
 
         // The proto percentiles are ordered ascending. We write the first match to the stream.
         double last_percentile = -1.;
-        for (const double p : {.0, .5, .75, .8, .9, .95, .99, .999, 1.}) {
+        for (const double p : {.5, .75, .8, .9, .95, .99, .999, 1.}) {
```
Comment (htuch) [inline source/client/output_collector_impl.cc]: Is this a bug fix?

File: test/integration/test_integration_basics.py [hunk 3]
```diff
@@ -271,3 +271,16 @@ def test_https_log_verbosity(https_test_server_fixture):
...
+def test_dotted_output_format(http_test_server_fixture):
+  """
+  Ensure we get the dotted string output format when requested.
+  and ensure we get latency percentiles.
+  """
+  output, _ = http_test_server_fixture.runNighthawkClient([
+      "--duration 1", "--rps 10", "--output-format dotted",
+      http_test_server_fixture.getTestServerRootUri()
+  ],
+                                                          as_json=False)
```
Comment (htuch) [inline test/integration/test_integration_basics.py]: This seems very strange formatting, are we running pyformat on the NH repo? See what Envoy does.

File: source/client/output_collector_impl.h [hunk 4]
```diff
@@ -50,5 +50,11 @@ class YamlOutputCollectorImpl : public OutputCollectorImpl {
   std::string toString() const override;
 };
 
+class DottedStringOutputCollectorImpl : public OutputCollectorImpl {
```
Comment (htuch) [inline source/client/output_collector_impl.h]: Can you link to where this format is specified? i.e. is there an RFC or something?

File: source/client/output_collector_impl.cc [hunk 5]
```diff
@@ -122,5 +122,51 @@ std::string YamlOutputCollectorImpl::toString() const {
...
+                        Envoy::Protobuf::util::TimeUtil::DurationToMicroseconds(statistic.mean()))
+         << std::endl;
+      ss << fmt::format("{}.pstdev: {}", prefix,
+                        Envoy::Protobuf::util::TimeUtil::DurationToMicroseconds(statistic.pstdev()))
+         << std::endl;
+
+      // The proto percentiles are ordered ascending. We write the first match to the stream.
+      // Revisit to see if we can factor out a common denominator with the CLI output formatter.
+      double last_percentile = -1.;
+      for (const double p : {.5, .75, .8, .9, .95, .99, .999, 1.}) {
```
Comment (htuch) [inline source/client/output_collector_impl.cc]: This iteration over the histogram seems like a common pattern shared by the different output formatters. Would it be possible to create a higher-order function or class that implements this, with the inner block supplied by a lambda?

File: source/client/output_collector_impl.cc (hunk 2 above)
Comment (oschaaf) [reply source/client/output_collector_impl.cc]: Yes, this is a bug-fix, forgot to mention it in the PR description. 
Should I file a separate issue on this?

File: test/integration/test_integration_basics.py (hunk 3 above)
Comment (oschaaf) [reply test/integration/test_integration_basics.py]: This is what `tools/check_format fix` outputs. I double checked it, because I also thought it looked rather strange. Maybe we're hitting some weird edge case?
And yup, we are running formatting checks in CI (and they are content with this).

File: source/client/output_collector_impl.h (hunk 4 above)
Comment (oschaaf) [reply source/client/output_collector_impl.h]: I'm actually not sure if there's a formal spec. @jplevyak would you know?

File: source/client/output_collector_impl.cc (hunk 5 above)
Comment (oschaaf) [reply source/client/output_collector_impl.cc]: Sure; can I take a TODO on that?

File: source/client/output_collector_impl.cc (hunk 2 above)
Comment (htuch) [reply source/client/output_collector_impl.cc]: It's fine, just best in future to split these out :)

File: test/integration/test_integration_basics.py (hunk 3 above)
Comment (htuch) [reply test/integration/test_integration_basics.py]: Fair enough, kind of strange though.

File: source/client/output_collector_impl.cc (hunk 5 above)
Comment (htuch) [reply source/client/output_collector_impl.cc]: I think it could be good to do in this PR, it should actually shrink the size of the PR.

File: test/test_data/output_collector.dotted.gold [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+worker_1.stat_id.pstdev: 816
+worker_1..samples: 0
+worker_1..mean: 0
+worker_1..pstdev: 0
+worker_1.bar:2
+worker_1.foo:1
+global.stat_id.samples: 3
+global.stat_id.mean: 2000
+global.stat_id.pstdev: 816
+global..samples: 0
```
Comment (jplevyak) [inline test/test_data/output_collector.dotted.gold]: What do the double periods mean? ".." ?

File: source/client/output_collector_impl.cc (hunk 2 above)
Comment (oschaaf) [reply source/client/output_collector_impl.cc]: Done in https://github.com/envoyproxy/nighthawk/pull/150/commits/3836ba1cc1e39f0265e7ec02714a6de417a4fb30

File: source/client/output_collector_impl.cc (hunk 5 above)
Comment (oschaaf) [reply source/client/output_collector_impl.cc]: Done in https://github.com/envoyproxy/nighthawk/pull/150/commits/3836ba1cc1e39f0265e7ec02714a6de417a4fb30

File: test/test_data/output_collector.dotted.gold (hunk 6 above)
Comment (oschaaf) [reply test/test_data/output_collector.dotted.gold]: re: `What do the double periods mean? ".." ?`
Sorry, forgot to get back to this. We're outputting a stat with an empty name.
We won't be seeing that in practice.

File: source/client/output_collector_impl.cc (hunk 2 above)
Comment (oschaaf) [reply source/client/output_collector_impl.cc]: So actually, not having it introduced a bug, we would miss the minimal value. Added it back.
